llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -4516,6 +4516,39 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
         }
       }
     }
+
+    // Optimize
+    //    (setcc (shift N00, N01C), 0, eq/ne) -> (setcc N00, 0, eq/ne)
+    // If all shifted out bits are known to be zero, then the zero'd ness
+    // doesn't change and we can omit the shift.
+    // If all shifted out bits are equal to at least one bit that isn't
+    // shifted out, then the zero'd ness doesn't change and we can omit the
+    // shift.
+    if ((Cond == ISD::SETEQ || Cond == ISD::SETNE) && C1.isZero() &&
+        N0.hasOneUse() &&
+        (N0.getOpcode() == ISD::SHL || N0.getOpcode() == ISD::SRL ||
+         N0.getOpcode() == ISD::SRA)) {
+      bool IsRightShift = N0.getOpcode() != ISD::SHL;
+      SDValue N00 = N0.getOperand(0);
+      // Quick checks based on exact/nuw/nsw flags.
+      if ((IsRightShift && N0->getFlags().hasExact()) ||
+          (!IsRightShift && N0->getFlags().hasNoUnsignedWrap()) ||
+          (!IsRightShift && N0->getFlags().hasNoSignedWrap()))
```

Comment:
Nit: this is just my personal preference to use `?:` in expressions like this.
```suggestion
      if (IsRightShift ? N0->getFlags().hasExact() :
          (N0->getFlags().hasNoUnsignedWrap() || N0->getFlags().hasNoSignedWrap()))
```